promql/promqltest/test.go

```diff
@@ -693,29 +792,77 @@ func (ev *evalCmd) expectMetric(pos int, m labels.Labels, vals ...parser.Sequenc
 	ev.expected[h] = entry{pos: pos, vals: vals}
 }
 
+// validateExpectedAnnotations validates expected messages and regex match actual annotations.
+func validateExpectedAnnotations(expr string, expectedAnnotations []expectCmd, actualAnnotations []string, line int, annotationType string) error {
+	if len(expectedAnnotations) == 0 {
+		return nil
+	}
+
+	if len(actualAnnotations) == 0 {
+		return fmt.Errorf("expected %s annotations but none were found for query %q (line %d)", annotationType, expr, line)
+	}
+
+	// Check if all expected annotations are found in actual.
+	for _, e := range expectedAnnotations {
+		matchFound := slices.ContainsFunc(actualAnnotations, e.CheckMatch)
+		if !matchFound {
+			return fmt.Errorf("expected %s annotation %q but no matching annotation was found for query %q (line %d)", annotationType, e.String(), expr, line)
```

Comment:
Might also be useful to include what annotations were returned by the query, rather than just saying no matching annotation was found - it may be that there's a typo in the pattern or the annotation message, or it's the wrong kind of annotation, and including the list of annotations (of both kinds) that were returned would make it easier to see.